r2ai/auto.py

```diff
@@ -407,7 +446,7 @@ def auto_chat_llama(interpreter):
         max_tokens=int(interpreter.env["llm.maxtokens"]),
         tools=tools,
         messages=interpreter.messages,
-        tool_choice="auto",
+        tool_choice="required",
```

Comment:
this won't work well, it means it'll only try to call a tool and never be able to respond with just text

Reply:
Just for llama or for everything else?

Reply:
We have to handle this case somehow then

__If a function is used, the output will contain "finish_reason": "tool_calls" in the response, as well as a tool_calls object that has the name of the function and the generated function arguments.__

Reply:
Issue is, no idea how to handle it, as it doesn't give any hint of what answer he expects back

Reply:
https://github.com/openai/openai-cookbook/blob/main/examples/How_to_call_functions_with_chat_models.ipynb

Reply:
I got it @dnakov 

We can keep it on required to avoid getting the finish reason, and give the assistans a tool they can use to reply with just text

Reply:
yeah I tried that a few months ago but auto worked a lot better

Reply:
I’ll wait to merge this, what should we do?

Reply:
No idea. From what I see it's based in the model. Using openrouter with claude has this issue, but when using openrouter with gpt-4 it continues to work just fine.

I guess switch back to auto and when finish reason is returned just print an error message

Reply:
![Screenshot_20240911-110701](https://github.com/user-attachments/assets/973641b5-c5bb-4a6b-9567-4e075a7f32f6)

Reply:
What about making it configurable with a default depending on the model?

Reply:
> What about making it configurable with a default depending on the model?

While logging some things I've noticed that the model gets confused about the functions, sometimes trying to run a radare2 command but without specifying that it's a radare2 command. I hope I can fix that with a new help function to have a model get supported r2 commands and distinguish them

Reply:
ahh come on

`{'function': {'arguments': '{"command": "aaa; iM; ie"}', 'name': None}, 'id': 'none', 'type': 'function'} `

Reply:
Or this works xD

![Screenshot from 2024-09-11 12-56-13](https://github.com/user-attachments/assets/ec3e4e0c-b6c8-484a-b840-6ee8e80220be)

Reply:
ah the arguments are none because we don't have required on them

Reply:
I think it's a good idea to implement [instructor](https://python.useinstructor.com) in the autoai instead of manually declaring everything. @dnakov @trufae what do you guys think?

Reply:
```
[09/11/24 14:35:37] DEBUG    r2ai - DEBUG - {'r2cmd': R2CmdResponse(command='aaa'), 'r2_help': '', 'python_cmd': '', 'raw': ''}                                                                           auto.py:158
                    DEBUG    r2ai - DEBUG - {'r2cmd': R2CmdResponse(command='iM'), 'r2_help': '', 'python_cmd': '', 'raw': ''}                                                                            auto.py:158
                    DEBUG    r2ai - DEBUG - {'r2cmd': R2CmdResponse(command='ie'), 'r2_help': '', 'python_cmd': '', 'raw': ''}                                                                            auto.py:158
                    DEBUG    r2ai - DEBUG - {'r2cmd': R2CmdResponse(command='i?'), 'r2_help': '', 'python_cmd': '', 'raw': ''}                                                                            auto.py:158
                    DEBUG    r2ai - DEBUG - {'r2cmd': R2CmdResponse(command='iz'), 'r2_help': '', 'python_cmd': '', 'raw': ''}                                                                            auto.py:158
                    DEBUG    r2ai - DEBUG - {'r2cmd': R2CmdResponse(command='iS'), 'r2_help': '', 'python_cmd': '', 'raw': ''} 
```

Reply:
I haven't seen this issue when using claude with litellm and auto, maybe its an openrouter thing?

Reply:
btw @nitanmarcel take a look at ui/chat.py (it's auto.py but using just litellm so it will support all the providers) in my UI PR. I'm thinking on just using litellm for mostly everything but have to figure out how to integrate it with llama.cpp so the web server works as well

Reply:
> btw @nitanmarcel take a look at ui/chat.py (it's auto.py but using just litellm so it will support all the providers) in my UI PR. I'm thinking on just using litellm for mostly everything but have to figure out how to integrate it with llama.cpp so the web server works as well

Will do, I'm currently implementing instructor so I don't loose my sanity parsing the openai json (see output above)

Reply:
You won't really need that with litellm though :)

Reply:
> You won't really need that with litellm though :)

yeah but will be easier with the others. I'm already rewriting the auto module

Reply:
Looks like I can't use streaming mode properly with instructor since it doesn't expose a way to get the use_tool_id to use it the way I want it while using models.

I'll put those back to auto tomorrow. And I still want to rewrite the assistant, but in a separate PR ^^